[ENH] Harmonize short repr of niimgs

NicolasGensollen opened the pull request:
Fixes #747 

**Description**

function `_repr_niimgs` now accepts a `shorten` boolean parameter to truncate display of niimgs (default is set to True). 
Truncated display was possible before (using `_utils.short_repr`) but not straightforward, and it was truncating on the repr rather than adapting to the type of data provided (see #747 for more details).

**Examples**

See also `test_repr_niimgs`

Previous behavior:

```python
from nilearn import _utils
print(_utils._repr_niimgs('/this/is/a/very/long/name/for/a/nifti/file'))
print("-"*80)
list_of_size_3 = ['/this/is/a/very/long/name/for/a/nifti/file',
                  '/this/is/another/very/long/name/for/a/nifti/file',
                  '/this/is/again/another/very/long/name/for/a/nifti/file']
print(_utils._repr_niimgs(list_of_size_3))
list_of_size_4 = ['/this/is/a/very/long/name/for/a/nifti/file',
                  '/this/is/another/very/long/name/for/a/nifti/file',
                  '/this/is/again/another/very/long/name/for/a/nifti/file',
                  '/this/is/again/another/super/very/long/name/for/a/nifti/file']
print("-"*80)
print(_utils._repr_niimgs(list_of_size_4))
```

Outputs:

```
/this/is/a/very/long/name/for/a/nifti/file
--------------------------------------------------------------------------------
[/this/is/a/very/long/name/for/a/nifti/file, /this/is/another/very/long/name/for/a/nifti/file, /this/is/again/another/very/long/name/for/a/nifti/file]
--------------------------------------------------------------------------------
[/this/is/a/very/long/name/for/a/nifti/file, /this/is/another/very/long/name/for/a/nifti/file, /this/is/again/another/very/long/name/for/a/nifti/file, /this/is/again/another/super/very/long/name/for/a/nifti/file]
```

Example of previous use of shortening:

```python
_utils.load_niimg(list_of_size_4)
```

Raises:

```
TypeError: Data given cannot be loaded because it is not compatible with nibabel format:
[/this/is/a/very/l...
```

New behavior

```python
from nilearn import _utils
print(_utils._repr_niimgs('/this/is/a/very/long/name/for/a/nifti/file'))
print("-"*80)
list_of_size_3 = ['/this/is/a/very/long/name/for/a/nifti/file',
                  '/this/is/another/very/long/name/for/a/nifti/file',
                  '/this/is/again/another/very/long/name/for/a/nifti/file']
print(_utils._repr_niimgs(list_of_size_3, shorten=True))
print("-"*80)
print(_utils._repr_niimgs(list_of_size_3, shorten=False))
list_of_size_4 = ['/this/is/a/very/long/name/for/a/nifti/file',
                  '/this/is/another/very/long/name/for/a/nifti/file',
                  '/this/is/again/another/very/long/name/for/a/nifti/file',
                  '/this/is/again/another/super/very/long/name/for/a/nifti/file']
print("-"*80)
print(_utils._repr_niimgs(list_of_size_4, shorten=True))
print("-"*80)
print(_utils._repr_niimgs(list_of_size_4, shorten=False))
```

Outputs:

```
/this/is/a/very/lo...
--------------------------------------------------------------------------------
[/this/is/a/very/lo..., /this/is/another/v..., /this/is/again/ano...]
--------------------------------------------------------------------------------
[/this/is/a/very/long/name/for/a/nifti/file, /this/is/another/very/long/name/for/a/nifti/file, /this/is/again/another/very/long/name/for/a/nifti/file]
--------------------------------------------------------------------------------
[/this/is/a/very/lo...,
         ...
 /this/is/again/ano...]
--------------------------------------------------------------------------------
[/this/is/a/very/long/name/for/a/nifti/file,
 /this/is/another/very/long/name/for/a/nifti/file,
 /this/is/again/another/very/long/name/for/a/nifti/file,
 /this/is/again/another/super/very/long/name/for/a/nifti/file]
```

```python
_utils.load_niimg(list_of_size_4)
```

```
TypeError: Data given cannot be loaded because it is not compatible with nibabel format:
[/this/is/a/very/lo...,
         ...
 /this/is/again/ano...]
```

Not sure I made the best choices for display, let me know if you have some modifications in mind.

bthirion commented on nilearn/_utils/niimg.py:
```diff
@@ -158,37 +158,65 @@ def copy_img(img):
         # Shorten the repr to have a useful error message
-        this_repr = this_repr[:18] + '...'
-    return this_repr
+        return niimg_rep[:(truncate-2)] + '...'
```
```suggestion
        return niimg_rep[: (truncate - 2)] + '...'
```

bthirion commented on nilearn/tests/test_niimg_conversions.py:
```diff
@@ -389,23 +389,88 @@ def test_iter_check_niimgs_memory():
+    assert (_utils._repr_niimgs('/this/is/a/very/long/name/for/a/nifti/file',
+                                shorten=False) ==
+            '/this/is/a/very/long/name/for/a/nifti/file')
+
```
Could be rewritten more nicely by using 
`long_str='this/is/a/very/long/name/for/a/nifti/file'`
`short_str = '/this/is/a/very/lo...'`

bthirion commented on nilearn/tests/test_niimg_conversions.py:
```diff
@@ -389,23 +389,88 @@ def test_iter_check_niimgs_memory():
+             " /this/is/another/very/long/name/for/a/nifti/file,\n"
+             " /this/is/again/another/very/long/name/for/a/nifti/file,\n"
+             " /this/is/again/another/super/very/long/name/for/a/nifti/file]"))
+
```
Can be written a bit mroe nicely too.

bthirion commented on nilearn/tests/test_niimg_conversions.py:
```diff
@@ -389,23 +389,88 @@ def test_iter_check_niimgs_memory():
+    # Shorten has no effect in this case
+    for shorten in [True, False]:
+        assert (
+            _utils._repr_niimgs(img1, shorten=shorten).replace("10L","10") ==
```
```suggestion
            _utils._repr_niimgs(img1, shorten=shorten).replace("10L", "10") ==
```

bthirion reviewed (commented):
LGTM ! 
I have only superficial comments.

jeromedockes commented:
looks good! I wonder if this will break doctests in libraries that depend on nilearn but given the places this function is used I'm guessing probably not. 

I think in most cases the end of the string (path) would be more informative than the beginning, ie "...a/subj1.nii.gz" rather than "/home/nilearn/dat..."

jeromedockes commented:
also should `pathlib.Path`s be converted to strings before further processing?

NicolasGensollen commented:
Thank you for your quick reviews and comments @bthirion @jeromedockes ! :+1: 

@bthirion : I added your suggestions and cleaned the test function

@jeromedockes : 

>I think in most cases the end of the string (path) would be more informative than the beginning, ie "...a/subj1.nii.gz" rather than "/home/nilearn/dat..."

I agree when it is a path. More precisely, I think we should display "...a/subj1.nii.gz"  when the path to the file is too long, and "sub1-super-long-nam..." when the name of the file itself is too long (or both are too long...). WDYT?
 
I implemented this behavior as well as `pathlib` support as you suggested (see last commits). 

Note: We could also take care of "too-long-path" case in pure string format by splitting on "/" and doing something similar as for `Path` objects. I can add this if you think it's worth it.

jeromedockes commented:
> I agree when it is a path. More precisely, I think we should display "...a/subj1.nii.gz"  when the path to the file is too long, and "sub1-super-long-nam..." when the name of the file itself is too long (or both are too long...). WDYT?
>  

sounds good! I think for simplicity I would have done in all cases: (i)
convert to string (ii) cut the beginning of the string if necessary; ie
something as basic as

```
# as before
length = max(length, 10)
file_path = str(file_path)
if len(file_path) <= length:
    return file_path
return "... {}".format(file_path[-(length - 4):])
```

but the behavior you implemented makes a lot of sense so we can keep it. 

> I implemented this behavior as well as `pathlib` support as you suggested (see last commits). 
>

great! a couple of details:

- to display paths correctly on windows I would replace `"{0}/{1}".format(p,
  rep)` with `str(Path(p, rep))` (or something equivalent)
- the current version will prepend "..." to a Path even when it is not
  truncated, and will add a "/" at the beginning if the path starts with "/", ie
  _repr_niimgs(Path("/a/b")) -> "...//a/b" (maybe avoiding to think about edge
  cases like this would be an argument for a more basic behavior such as the one
  above)

> Note: We could also take care of "too-long-path" case in pure string format by splitting on "/" and doing something similar as for `Path` objects. I can add this if you think it's worth it.

splitting on "/" will not work on windows, but when input is a string it
will always represent a path so we can convert to Path so that all paths
are treated consistently, whether provided as a string (which will be
the most common case) or Path

NicolasGensollen commented:
Thanks for the comments @jeromedockes !
I added your suggestions and simplified the code a little bit. Let me know if you think that's good enough to merge.

bthirion reviewed (approved):
LGTM, thx !

jeromedockes reviewed (approved):
LGTM!

NicolasGensollen commented:
Merging... Thanks!
